### DISCUSSION TITLE
in next.js 14 how to pass data from one page to another without showing data on url?

### DISCUSSION THREAD

**Discussion by xts-bit** [vercel/next.js#57756] [Help]
### Summary

I am a React Native dev trying to build a web app my quastion is in next.js 14 how to pass data from one page to another without showing data on url? I have an app directory in my Next.js project.

by not showing data on URL i mean for example http://localhost:3000/mypage?name="Johndeo" 

I have a Button component that I am using in my whole app for Button. it also has a href prop for navigating from one page to another. 

i want to access pass name as data from page1 and log it to page2

### Additional information

```js
This is how i am using Button component
`<Button href="/mypage2">     Go to page2 </Button>`
```
Buttom component:

```
import React from 'react';
import Link from 'next/link'
import clsx from 'clsx'

interface ButtonProps {
  type?: "button" | "submit" | "reset";
  href?: string;
  className?: string;
  children: React.ReactNode;
}

export function Button({ href, className, children, ...props }: ButtonProps) {

  className = clsx(
    className,
    'inline-flex rounded-full px-4 py-1.5 text-sm font-semibold transition bg-white text-black hover:bg-neutral-200'
  )

  let inner = <span className="relative top-px">{children}</span>

  if (href) {
    return (
      <Link href={href} className={className} {...props}>
        {inner}
      </Link>
    )
  }

  return (
    <button className={className} {...props}>
      {inner}
    </button>
  )
}
```

**Comment by yardenhochman** [comment]
I'm new to NextJS but I'll try to answer.
On the web, sharing data across multiple pages can be performed via the url as you mentioned, and also via cookies, and then JS (state management). Lastly, you can also save the data on a DB and fetch it based on the user-id any time you need it.
The best option really depends on the use case.


the viewing user id, for example, is session related, and not page related. so normally that would be stored in a cookie.
the current context of the page, for example, the profile of some other user, will be in the url localhost:3000/profile/john-doe

Finally, as explained in the last option I gave, you could potentially save the name of the user in some db, and then fetch it in whatever page you please based on the user-id...
This might be the cleanest option in many cases since the source of truth is persistent, and also, cached automatically by nextjs's fetch implementation

**Reply by meabhipatel** [reply]
Absolutely these are the ways to pass data from one to another but in react using react router dom package we can share data between two pages. which can be help full in many scenarios. I am also finding the solution for my use cases

**Comment by rachid-serraf** [comment]
Use a client-side solution like Context or localStorage.
Next.js pages can’t pass hidden data through <Link>, so store the value in Context/localStorage on page1 and read it on page2.
